misc/python/materialize/data_ingest/definition.py

```diff
@@ -0,0 +1,185 @@
+            transactions.append(
+                Transaction(
+                    [
+                        RowList(
+                            [
+                                Row(
+                                    fields=deepcopy(fields_with_values),
+                                    operation=Operation.INSERT,
+                                )
+                            ]
+                        )
+                    ]
+                )
+            )
+
+        return transactions
+
+
+class Upsert(Definition):
+    def __init__(self, keyspace: Keyspace, count: Records, record_size: RecordSize):
+        self.keyspace = keyspace
+        self.count = count
+        self.record_size = record_size
+
+    def generate(self, fields: List[Field]) -> List[Transaction]:
+        if self.count == Records.ONE:
+            count = 1
+        elif self.count == Records.MANY:
+            count = 1000
+        else:
+            raise ValueError(f"Unexpected count {self.count}")
+
+        fields_with_values = deepcopy(fields)
+
+        transactions = []
+        for i in range(count):
+            for field in fields_with_values:
+                if field.is_key:
+                    if self.keyspace == Keyspace.SINGLE_VALUE:
+                        field.value = field.typ.num_value(0)
+                    else:
+                        raise NotImplementedError
+                else:
+                    field.value = field.typ.random_value(self.record_size)
+
+            transactions.append(
+                Transaction(
+                    [
+                        RowList(
+                            [
+                                Row(
+                                    fields=deepcopy(fields_with_values),
+                                    operation=Operation.UPSERT,
+                                )
+                            ]
+                        )
+                    ]
+                )
+            )
+
+        return transactions
+
+
+class Delete(Definition):
+    def __init__(
+        self,
+        number_of_records: Records,
+        record_size: RecordSize,
+        num: Optional[int] = None,
+    ):
+        self.number_of_records = number_of_records
+        self.record_size = record_size
+        self.num = num
+
+    def generate(self, fields: List[Field]) -> List[Transaction]:
+        transactions = []
+
+        fields_with_values = [field for field in fields if field.is_key]
+
+        if self.number_of_records == Records.ONE:
+            for field in fields_with_values:
+                field.value = field.typ.random_value(self.record_size)
+            transactions.append(
+                Transaction([RowList([Row(fields_with_values, Operation.DELETE)])])
+            )
+        elif self.number_of_records == Records.MANY:
+            for i in range(1000):
+                for field in fields_with_values:
+                    field.value = field.typ.random_value(self.record_size)
```

Comment:
Is it ok / desired to delete a row multiple times (if random picks the same value multiple times)?

Reply:
Yes